File: ts/examples/quickstart.ts
// Run from ts/: npm run quickstart

import { MemoryStorage, SessionKnit } from '../src/index.js'

interface Message {
  role: string
  content: string
  hasToolCall?: boolean
}

const storage = new MemoryStorage<Message>()
const sessionknit = new SessionKnit<Message>(storage, {
  hasUnresolvedToolCall: (m) => m.hasToolCall === true,
  buildContinuation: (m) => ({
    role: 'user',
    content: `[resumed after interruption during: ${m.content}]`,
  }),
})

// --- 1. Topology repair: a turn with two parallel tool calls ---
const sessionId = 'demo-session'

await sessionknit.append(sessionId, { id: 'a', parentId: null, message: { role: 'user', content: 'Summarize a.txt and b.txt.' } })
await sessionknit.append(sessionId, { id: 'b', parentId: 'a', message: { role: 'assistant', content: 'Reading both files.' } })
// Two parallel tool results — both children of 'b', siblings of each other.
await sessionknit.append(sessionId, { id: 'tool1', parentId: 'b', message: { role: 'tool', content: 'a.txt: revenue grew 12%' } })
await sessionknit.append(sessionId, { id: 'tool2', parentId: 'b', message: { role: 'tool', content: 'b.txt: tickets dropped 8%' } })
// The next turn only descends from ONE of them — a naive walk from 'd'
// would never see tool1.
await sessionknit.append(sessionId, { id: 'd', parentId: 'tool2', message: { role: 'assistant', content: 'Both summarized.' } })

console.log('--- topology repair ---')
const repaired = await sessionknit.resume(sessionId, 'd')
console.log('reconstructed order:', repaired.messages.map((m) => m.content))
console.log(
  'includes the dropped sibling (tool1)?',
  repaired.messages.some((m) => m.content.includes('revenue')),
)

// --- 2. Interruption recovery ---
const crashSession = 'crash-session'
await sessionknit.append(crashSession, { id: 'x', parentId: null, message: { role: 'user', content: 'Write a summary to disk.' } })
await sessionknit.append(crashSession, {
  id: 'y',
  parentId: 'x',
  message: { role: 'assistant', content: 'Calling write_file...', hasToolCall: true },
})
// Process "crashed" here — no tool result was ever recorded.

console.log('\n--- interruption recovery ---')
const resumed = await sessionknit.resume(crashSession)
console.log('resumed after interruption:', resumed.resumedAfterInterruption)
console.log('final messages:', resumed.messages.map((m) => m.content))

File: py/examples/quickstart.py
"""Run from py/: PYTHONPATH=src python3 examples/quickstart.py"""

import asyncio
from dataclasses import dataclass

from sessionknit import MemoryStorage, SessionKnit, SessionEntry


@dataclass
class Message:
    role: str
    content: str
    has_tool_call: bool = False


async def main() -> None:
    storage = MemoryStorage()
    sessionknit = SessionKnit(
        storage,
        has_unresolved_tool_call=lambda m: m.has_tool_call,
        build_continuation=lambda m: Message(
            role="user", content=f"[resumed after interruption during: {m.content}]"
        ),
    )

    # --- 1. Topology repair: a turn with two parallel tool calls ---
    session_id = "demo-session"

    await sessionknit.append(
        session_id,
        SessionEntry(id="a", parent_id=None, message=Message(role="user", content="Summarize a.txt and b.txt.")),
    )
    await sessionknit.append(
        session_id,
        SessionEntry(id="b", parent_id="a", message=Message(role="assistant", content="Reading both files.")),
    )
    # Two parallel tool results — both children of 'b', siblings of each other.
    await sessionknit.append(
        session_id,
        SessionEntry(id="tool1", parent_id="b", message=Message(role="tool", content="a.txt: revenue grew 12%")),
    )
    await sessionknit.append(
        session_id,
        SessionEntry(id="tool2", parent_id="b", message=Message(role="tool", content="b.txt: tickets dropped 8%")),
    )
    # The next turn only descends from ONE of them — a naive walk from
    # 'd' would never see tool1.
    await sessionknit.append(
        session_id,
        SessionEntry(id="d", parent_id="tool2", message=Message(role="assistant", content="Both summarized.")),
    )

    print("--- topology repair ---")
    repaired = await sessionknit.resume(session_id, "d")
    print("reconstructed order:", [m.content for m in repaired.messages])
    print(
        "includes the dropped sibling (tool1)?",
        any("revenue" in m.content for m in repaired.messages),
    )

    # --- 2. Interruption recovery ---
    crash_session = "crash-session"
    await sessionknit.append(
        crash_session,
        SessionEntry(id="x", parent_id=None, message=Message(role="user", content="Write a summary to disk.")),
    )
    await sessionknit.append(
        crash_session,
        SessionEntry(
            id="y",
            parent_id="x",
            message=Message(role="assistant", content="Calling write_file...", has_tool_call=True),
        ),
    )
    # Process "crashed" here — no tool result was ever recorded.

    print("\n--- interruption recovery ---")
    resumed = await sessionknit.resume(crash_session)
    print("resumed after interruption:", resumed.resumed_after_interruption)
    print("final messages:", [m.content for m in resumed.messages])


if __name__ == "__main__":
    asyncio.run(main())
